salt/orch/force_removal.sls

```diff
@@ -0,0 +1,72 @@
+# This is a best effort orchestration. Any operation run on the minions by this orchestration
+# could fail, as we don't know the current state of the cluster at this point. The idea is to
+# try to clean as much as possible where possible, but in the end the only thing we can really
+# ensure is that the salt-key of the target is rejected.
+
+# must provide the node (id) to be removed in the 'target' pillar
+{%- set target = salt['pillar.get']('target') %}
+{%- set target_nodename = salt.saltutil.runner('mine.get', tgt=target, fun='nodename')[target] %}
+# services that will be called when trying to unregister as clean as possible
+{%- set services_to_stop = { "kubelet": ["stop", "remove-pre-reboot"], "etcd": ["remove-pre-stop-services", "remove-pre-reboot"] } %}
+# nodes that are responding
+{%- set up_nodes = salt.saltutil.runner('manage.up', tgt='P@roles:kube-* and not ' + target, expr_form='compound') %}
+
+# Try to unregister the machine from cluster services. It's important to keep every sls in
+# different toplevel stages, so each one will be tried even if previous fails for whatever
+# reason.
+{%- for service, sls_list in services_to_stop.items() %}
+  {%- for sls in sls_list %}
+stop-{{ service }}-{{ sls }}-stage:
+  salt.state:
+    - tgt: {{ target }}
+    - fail_minions: {{ target }}
+    - sls:
+        - {{ service }}.{{ sls }}
+  {%- endfor %}
+{%- endfor %}
+
+# Try to shutdown the target
+shutdown-target:
+  salt.function:
+    - tgt: {{ target }}
+    - name: cmd.run
+    - fail_minions: {{ target }}
+    - arg:
+      - systemctl poweroff
+    - kwarg:
+        bg: True
+
+# Even if trying to unregister the machine failed (e.g. `kubectl delete node` failed because
+# kubelet/stop.sls couldn't be executed on the target), we try to remove the node again from
+# the kubernetes cluster. This time from other working machine, and after we have done our
+# best to shut down the target. Do this from all the machines we can, best effort again.
+# Once that we managed to remove the node from the cluster, we stop.
+{%- set previous_node = none %}
+{%- for node in up_nodes %}
+try-to-unregister-{{ target }}-from-{{ node }}:
+  salt.function:
+    - tgt: {{ node }}
+    - name: cmd.run
+    - fail_minions: {{ node }}
+    - arg:
+        - kubectl --kubeconfig={{ pillar['paths']['kubeconfig'] }} delete node {{ target_nodename }}
+  {%- if previous_node %}
+    - onfail:
+        - try-to-unregister-{{ target }}-from-{{ previous_node }}
+  {%- endif %}
+  {%- set previous_node = node %}
+{%- endfor %}
```

Comment:
Why don't you just run the `kubectl delete node` in a master that is up and running? Get all the masters that are up and running (with the `mine.get`, targeting `G@roles:kube-master and not L@node_down1,...`), and then running this salt function on `tgt: {{ responsive_masters|first }}`

Reply:
Because we can't assume that a master that has salt-minion responding will succeed with running `kubectl delete node`.

Reply:
If no masters have a Salt minion responding you better reinstall your cluster... :-)

Anyway, you can get any responsive machine in the cluster where `kubectl` can be run (a master or a minion), and just run it there instead of looping.

Reply:
There's a long way between "no masters have a salt minion responding" and "I will just try with the first I got". Besides, my whole point is that the master you choose might have salt-minion responding, but `kubectl delete node` could fail there.

Please note that in this case as long as one succeeds running `kubectl delete node` the rest afterwards won't be executed because of the `onfail` requisite.

Reply:
We must be prepared for the target not to respond, but we don't have to be prepared for "maybe all machines in the cluster will fail to run any command" because, if that is the case, you better reinstall the cluster...

Reply:
As stated several times previously, this is a last resort orchestration. You could have two failing machines, you can only remove one by one, and it could happen that `masters|first` is the other machine failed. As a last resort orchestration I think this is the right way to go, go best effort everywhere you can.

Reply:
Yes, I could have a deployment with say, 5 masters. I don't follow your reasoning here. 5 masters, 40 workers. 1 unresponsive master and 1 unresponsive worker. I click to force remove the worker, and `masters|first` is the unresponsive master...

Reply:
By the way, you **can** add new masters.

Reply:
It is not `masters|first` but `responsive_masters|first`, and

```SaltStack
# get nodes_down with manage.down()
{% responsive_masters = salt['mine.get'](
    'G@roles:kube-master and not L@' + nodes_down|join(','),
    'nodename', 'compound').keys() %}
```

so those masters will be responsive (at least for Salt).

Reply:
I have stated several times before that a minion responding for salt does not mean anything more than its `salt-minion` is working and talking to the master. You don't know anything about that machine and if `kubectl delete node` will work or not.

Reply:
... and you needt least a kubernetes master with a salt minion that is responsive _and_ a working `kubectl` or otherwise your cluster is utterly broken and you will not be able to do anything. All our orchestrations depend on that, so if you remove an unresponsive node and you only leave a unresponsive master, the cluster is unsable.

Reply:
Let's try to keep this code as simple as possible. Let's go to the first responsive master and issue the `kubectl delete` statement. If that proves to be to fragile we can always bring this code back.